capstone-backend/src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -81,27 +94,42 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
       }
       BufferedReader br = new BufferedReader(streamReader);
 
-      String output;
-      while ((output = br.readLine()) != null) {
-        fullOutput += output;
-      }
+      fullOutput = br.lines().collect(Collectors.joining());
 
       if (conn.getResponseCode() != 200) {
         System.err.println(fullOutput);
-        throw new RuntimeException("Failed : error code : " + conn.getResponseCode());
+        response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+        return;
       }
 
       conn.disconnect();
-    } catch (MalformedURLException e) {
-      e.printStackTrace();
-    } catch (IOException e) {
+    } catch (Exception e) {
       e.printStackTrace();
+      response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+      return;
     }
 
     response.setContentType("application/json;");
-    Gson gson = new GsonBuilder().disableHtmlEscaping().create();
 
-    response.getWriter().println(gson.toJson(convertResponseToVolumeData(fullOutput)));
+    Collection<VolumeData> volumes = new ArrayList<>();
+    if (request.getParameter("gbookId") != null) {
+      VolumeData bookObject = individualBookToVolumeData(JsonParser.parseString(fullOutput));
+      if (bookObject != null) {
+        volumes.add(bookObject);
+      } else {
+        response.sendError(HttpServletResponse.SC_NOT_FOUND);
+        return;
+      }
+    } else if (request.getParameter("searchTerm") != null) {
+      volumes.addAll(convertResponseToVolumeData(fullOutput));
+      if (volumes.size() == 0) {
+        response.sendError(HttpServletResponse.SC_NOT_FOUND);
+        return;
+      }
+    }
+
+    Gson gson = new GsonBuilder().disableHtmlEscaping().create();
```

Comment:
Worth looking into if this creates any security concerns. I doubt it will as all the data comes from the Google Books API, but disabling HTML escaping always seems a little risky.

Reply:
Without disabling HtmlEscaping, strings like `\u003d` show up in place of other special characters like `=` or `&`. I think we can trust the Google Books API to return safe responses.